[REVIEW] make gather_bitmask_kernal more reusable.

cwharris opened the pull request:
I need functionality similar to `gather_bitmask_kernel` for stack + tile #3403, but I need the ability to select validity from multiple columns. For instance, `stack` (now called `interleave_columns`) takes multiple columns and interleaves them to produce a new column. As it stands, `gather_bitmask_kernel` cannot reshape multiple columns in to a single column.

This PR refactors `gather_bitmask_kernel` to accept a validity selector, which is able to arbitrarily provide validity for any given bit. There is an example validity selector `gather_bitmask_functor` which extracts the "ignore out of bounds" functionality from the original kernel.

jrhemstad commented on cpp/include/cudf/detail/gather.cuh:
```diff
@@ -45,57 +45,86 @@ struct bounds_checker {
 };
 
+template<typename ValiditySelector>
+__device__ void select_bitmask(ValiditySelector get_validity,
```
Needs docs.

jrhemstad commented on cpp/include/cudf/detail/gather.cuh:
```diff
@@ -46,56 +46,104 @@ struct bounds_checker {
+                                      size_type mask_count,
+                                      size_type mask_num_bits,
+                                      size_type* validity_counts) {
+  select_bitmask(get_validity, masks, mask_count, mask_num_bits, validity_counts);
```
It's a bit unusual to have a kernel that just calls a device function. Can the body of the function just be moved into the kernel? 

I suspect this was done to preserve the existing `gather_bitmask_kernel`, but that can just be factored out to instead call the new kernel.

cwharris replied:
Sounds good to me! I wasn't sure how much we wanted to change in a single PR. Right now `scatter` and `gather` are the only functions using the gather kernel. I'll remove `__global__ void gather_bitmask_kernel` and move the Selector initialization up a level.

trevorsm7 commented on cpp/include/cudf/detail/gather.cuh:
```diff
@@ -46,56 +46,104 @@ struct bounds_checker {
-        const bool thread_active = destination_row < destination_table_num_rows;
-        size_type source_row =
-          thread_active ? gather_map[destination_row] : 0;
+      if (threadIdx.x % warp_size == 0) {
```
```suggestion
      if (threadIdx.x % warp_size == 0 && bit_active) {
```
If `threadIdx.x >= warp_size`, wouldn't it be possible for `bit_idx >= mask_num_bits`?

cwharris replied:
Good catch! Thanks!

cwharris commented on cpp/include/cudf/detail/gather.cuh:
```diff
@@ -45,57 +45,86 @@ struct bounds_checker {
-        } else {
-          bit_is_valid = source_col.is_valid(source_row);
-        }
+      validity_count_aggregate += single_lane_block_popc_reduce(warp_validity);
```
> not all threads will enter the grid stride loop

```
auto bit_idx_offset = blockIdx.x * blockDim.x;
auto validity_count_aggregate = static_cast<size_type>(0);
while (bit_idx_offset < mask_num_bits) {
    auto const bit_idx = bit_idx_offset + threadIdx.x;
    ...
    validity_count_aggregate += single_lane_block_popc_reduce(warp_validity);
    bit_idx_offset += blockDim.x * gridDim.x;
}
```

It's _very_ possible I don't know what I'm talking about, but this seems to be written such that every thread _in a block_ enters this loop and calls `single_lane_block_popc_reduce`. It's possible that some blocks (and therefore all threads in that block) do not enter the loop. Do all threads _in a grid_ need to enter the loop?

jrhemstad replied:
Ah, you are correct, I missed how `bit_idx_offset` was being initialized. I thought it was unique per thread, but it is unique per block. So there are no correctness issues here.

However, my latter point about efficiency still remains. It's better to just do the accumulation in the register in the `while` and then do the warp reduction outside the loop.

jrhemstad commented on cpp/include/cudf/detail/gather.cuh:
```diff
@@ -46,58 +46,82 @@ struct bounds_checker {
+        mask[validity_idx] = warp_validity;
       }
+
+      validity_count_aggregate += single_lane_block_popc_reduce(warp_validity);
```
I didn't notice before, but this is the older version of this utility. Here is the newer, more general version which should be used instead: https://github.com/rapidsai/cudf/blob/branch-0.12/cpp/include/cudf/detail/utilities/cuda.cuh#L87

cwharris commented:
I turned off turbo boost and set my min processor to 100% while running these.
I threw the results in a spreadsheet and just divided the refactor branch by branch-0.12 and ended up with these as the averages (verifying the units were the same for each cell):
```
             refactor/control
Time:             100.00%
CPU:               99.93%
Iterations:        99.77%
bytes_per_second: 100.02%
```

branch 0.12 running on my machine.
```
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
----------------------------------------------------------------------------------------------------------
Benchmark                                                Time             CPU   Iterations UserCounters...
----------------------------------------------------------------------------------------------------------
Gather/double_coalesce_x/1024/1/manual_time          32282 ns        53639 ns        21817 bytes_per_second=484.009M/s
Gather/double_coalesce_x/2048/1/manual_time          31957 ns        53379 ns        21653 bytes_per_second=977.869M/s
Gather/double_coalesce_x/4096/1/manual_time          32484 ns        53874 ns        21480 bytes_per_second=1.87892G/s
Gather/double_coalesce_x/8192/1/manual_time          32573 ns        53775 ns        21400 bytes_per_second=3.74761G/s
Gather/double_coalesce_x/16384/1/manual_time         33170 ns        54362 ns        20829 bytes_per_second=7.36019G/s
Gather/double_coalesce_x/32768/1/manual_time         33207 ns        54263 ns        20990 bytes_per_second=14.704G/s
Gather/double_coalesce_x/65536/1/manual_time         34281 ns        55330 ns        20265 bytes_per_second=28.4871G/s
Gather/double_coalesce_x/131072/1/manual_time        43406 ns        64108 ns        15970 bytes_per_second=44.9964G/s
Gather/double_coalesce_x/262144/1/manual_time        50267 ns        71590 ns        13788 bytes_per_second=77.7097G/s
Gather/double_coalesce_x/524288/1/manual_time        60080 ns        80760 ns        11567 bytes_per_second=130.035G/s
Gather/double_coalesce_x/1048576/1/manual_time       90914 ns       106239 ns         7834 bytes_per_second=171.865G/s
Gather/double_coalesce_x/2097152/1/manual_time      142542 ns       157909 ns         4904 bytes_per_second=219.234G/s
Gather/double_coalesce_x/4194304/1/manual_time      248801 ns       264334 ns         2812 bytes_per_second=251.205G/s
Gather/double_coalesce_x/8388608/1/manual_time      456107 ns       471883 ns         1535 bytes_per_second=274.058G/s
Gather/double_coalesce_x/16777216/1/manual_time     873077 ns       889549 ns          801 bytes_per_second=286.344G/s
Gather/double_coalesce_x/33554432/1/manual_time    1708737 ns      1727031 ns          410 bytes_per_second=292.614G/s
Gather/double_coalesce_x/67108864/1/manual_time    3394309 ns      3356511 ns          207 bytes_per_second=294.611G/s
Gather/double_coalesce_x/1024/2/manual_time          37253 ns        58864 ns        17889 bytes_per_second=838.857M/s
Gather/double_coalesce_x/2048/2/manual_time          37757 ns        59239 ns        18160 bytes_per_second=1.61654G/s
Gather/double_coalesce_x/4096/2/manual_time          37889 ns        59191 ns        17866 bytes_per_second=3.2218G/s
Gather/double_coalesce_x/8192/2/manual_time          43976 ns        65328 ns        15903 bytes_per_second=5.55166G/s
Gather/double_coalesce_x/16384/2/manual_time         44432 ns        65834 ns        15566 bytes_per_second=10.9893G/s
Gather/double_coalesce_x/32768/2/manual_time         45032 ns        66116 ns        15605 bytes_per_second=21.6861G/s
Gather/double_coalesce_x/65536/2/manual_time         46787 ns        68087 ns        15028 bytes_per_second=41.7449G/s
Gather/double_coalesce_x/131072/2/manual_time        51536 ns        72770 ns        13501 bytes_per_second=75.7958G/s
Gather/double_coalesce_x/262144/2/manual_time        61306 ns        82600 ns        11233 bytes_per_second=127.435G/s
Gather/double_coalesce_x/524288/2/manual_time        88444 ns       109059 ns         7588 bytes_per_second=176.665G/s
Gather/double_coalesce_x/1048576/2/manual_time      148020 ns       160687 ns         4755 bytes_per_second=211.12G/s
Gather/double_coalesce_x/2097152/2/manual_time      252859 ns       268232 ns         2763 bytes_per_second=247.173G/s
Gather/double_coalesce_x/4194304/2/manual_time      461096 ns       477062 ns         1518 bytes_per_second=271.093G/s
Gather/double_coalesce_x/8388608/2/manual_time      880256 ns       896772 ns          792 bytes_per_second=284.008G/s
Gather/double_coalesce_x/16777216/2/manual_time    1716448 ns      1735279 ns          408 bytes_per_second=291.299G/s
Gather/double_coalesce_x/33554432/2/manual_time    3385954 ns      3411473 ns          207 bytes_per_second=295.338G/s
Gather/double_coalesce_x/67108864/2/manual_time    6726725 ns      6780940 ns          104 bytes_per_second=297.322G/s
Gather/double_coalesce_x/1024/4/manual_time          60095 ns        81559 ns        11438 bytes_per_second=1040.03M/s
Gather/double_coalesce_x/2048/4/manual_time          60633 ns        81050 ns        10486 bytes_per_second=2.01325G/s
Gather/double_coalesce_x/4096/4/manual_time          60846 ns        81730 ns        11306 bytes_per_second=4.01241G/s
Gather/double_coalesce_x/8192/4/manual_time          59756 ns        81286 ns        11250 bytes_per_second=8.17122G/s
Gather/double_coalesce_x/16384/4/manual_time         60156 ns        80929 ns        11268 bytes_per_second=16.234G/s
Gather/double_coalesce_x/32768/4/manual_time         60363 ns        80994 ns        11119 bytes_per_second=32.3565G/s
Gather/double_coalesce_x/65536/4/manual_time         60247 ns        81746 ns        11065 bytes_per_second=64.837G/s
Gather/double_coalesce_x/131072/4/manual_time        64908 ns        85516 ns        10734 bytes_per_second=120.362G/s
Gather/double_coalesce_x/262144/4/manual_time        92581 ns       112965 ns         7472 bytes_per_second=168.771G/s
Gather/double_coalesce_x/524288/4/manual_time       148241 ns       168169 ns         4516 bytes_per_second=210.806G/s
Gather/double_coalesce_x/1048576/4/manual_time      254208 ns       268467 ns         2781 bytes_per_second=245.862G/s
Gather/double_coalesce_x/2097152/4/manual_time      472384 ns       488228 ns         1484 bytes_per_second=264.615G/s
Gather/double_coalesce_x/4194304/4/manual_time      888928 ns       905496 ns          787 bytes_per_second=281.238G/s
Gather/double_coalesce_x/8388608/4/manual_time     1724841 ns      1743133 ns          405 bytes_per_second=289.882G/s
Gather/double_coalesce_x/16777216/4/manual_time    3402505 ns      3428842 ns          206 bytes_per_second=293.901G/s
Gather/double_coalesce_x/33554432/4/manual_time    6745587 ns      6800150 ns          104 bytes_per_second=296.49G/s
Gather/double_coalesce_x/67108864/4/manual_time   13440386 ns     13610351 ns           52 bytes_per_second=297.611G/s
Gather/double_coalesce_x/1024/8/manual_time          87176 ns       107905 ns         7882 bytes_per_second=1.40027G/s
Gather/double_coalesce_x/2048/8/manual_time          86732 ns       107186 ns         7871 bytes_per_second=2.81489G/s
Gather/double_coalesce_x/4096/8/manual_time          87429 ns       107881 ns         7848 bytes_per_second=5.58486G/s
Gather/double_coalesce_x/8192/8/manual_time          87292 ns       107425 ns         7890 bytes_per_second=11.1873G/s
Gather/double_coalesce_x/16384/8/manual_time         87579 ns       107668 ns         7837 bytes_per_second=22.3014G/s
Gather/double_coalesce_x/32768/8/manual_time         87174 ns       107377 ns         7337 bytes_per_second=44.8098G/s
Gather/double_coalesce_x/65536/8/manual_time         87462 ns       107636 ns         7806 bytes_per_second=89.3248G/s
Gather/double_coalesce_x/131072/8/manual_time        99767 ns       119939 ns         6984 bytes_per_second=156.615G/s
Gather/double_coalesce_x/262144/8/manual_time       157504 ns       176753 ns         4475 bytes_per_second=198.408G/s
Gather/double_coalesce_x/524288/8/manual_time       262551 ns       281803 ns         2665 bytes_per_second=238.049G/s
Gather/double_coalesce_x/1048576/8/manual_time      468032 ns       485757 ns         1484 bytes_per_second=267.075G/s
Gather/double_coalesce_x/2097152/8/manual_time      908496 ns       925281 ns          771 bytes_per_second=275.18G/s
Gather/double_coalesce_x/4194304/8/manual_time     1741044 ns      1759939 ns          402 bytes_per_second=287.184G/s
Gather/double_coalesce_x/8388608/8/manual_time     3414607 ns      3440573 ns          205 bytes_per_second=292.859G/s
Gather/double_coalesce_x/16777216/8/manual_time    6771298 ns      6827741 ns          103 bytes_per_second=295.364G/s
Gather/double_coalesce_x/33554432/8/manual_time   13454171 ns     13623636 ns           52 bytes_per_second=297.306G/s
Gather/double_coalesce_x/67108864/8/manual_time   26850495 ns     27481826 ns           26 bytes_per_second=297.946G/s
Gather/double_coalesce_o/1024/1/manual_time          32579 ns        54002 ns        21285 bytes_per_second=479.602M/s
Gather/double_coalesce_o/2048/1/manual_time          33384 ns        54195 ns        21134 bytes_per_second=936.08M/s
Gather/double_coalesce_o/4096/1/manual_time          32907 ns        54190 ns        21309 bytes_per_second=1.85479G/s
Gather/double_coalesce_o/8192/1/manual_time          32972 ns        53977 ns        21336 bytes_per_second=3.70224G/s
Gather/double_coalesce_o/16384/1/manual_time         33552 ns        54671 ns        20754 bytes_per_second=7.27659G/s
Gather/double_coalesce_o/32768/1/manual_time         34404 ns        55321 ns        20416 bytes_per_second=14.1926G/s
Gather/double_coalesce_o/65536/1/manual_time         40033 ns        60913 ns        17346 bytes_per_second=24.394G/s
Gather/double_coalesce_o/131072/1/manual_time        46317 ns        67053 ns        15076 bytes_per_second=42.1682G/s
Gather/double_coalesce_o/262144/1/manual_time        52365 ns        73225 ns        13344 bytes_per_second=74.5971G/s
Gather/double_coalesce_o/524288/1/manual_time        77233 ns        93834 ns         9030 bytes_per_second=101.154G/s
Gather/double_coalesce_o/1048576/1/manual_time      160252 ns       175455 ns         4353 bytes_per_second=97.5025G/s
Gather/double_coalesce_o/2097152/1/manual_time      522180 ns       537671 ns         1333 bytes_per_second=59.8453G/s
Gather/double_coalesce_o/4194304/1/manual_time     1215861 ns      1232259 ns          577 bytes_per_second=51.4039G/s
Gather/double_coalesce_o/8388608/1/manual_time     2648929 ns      2666681 ns          264 bytes_per_second=47.1889G/s
Gather/double_coalesce_o/16777216/1/manual_time    5516562 ns      5536903 ns          127 bytes_per_second=45.3181G/s
Gather/double_coalesce_o/33554432/1/manual_time   11309234 ns     11342638 ns           61 bytes_per_second=44.2117G/s
Gather/double_coalesce_o/67108864/1/manual_time   26227155 ns     26318070 ns           27 bytes_per_second=38.1284G/s
Gather/double_coalesce_o/1024/2/manual_time          37521 ns        59231 ns        18066 bytes_per_second=832.862M/s
Gather/double_coalesce_o/2048/2/manual_time          37968 ns        59681 ns        18467 bytes_per_second=1.60754G/s
Gather/double_coalesce_o/4096/2/manual_time          37960 ns        59153 ns        18073 bytes_per_second=3.21574G/s
Gather/double_coalesce_o/8192/2/manual_time          43993 ns        65322 ns        15983 bytes_per_second=5.54953G/s
Gather/double_coalesce_o/16384/2/manual_time         45531 ns        66869 ns        15363 bytes_per_second=10.7242G/s
Gather/double_coalesce_o/32768/2/manual_time         46458 ns        67592 ns        14527 bytes_per_second=21.0205G/s
Gather/double_coalesce_o/65536/2/manual_time         48376 ns        69767 ns        13509 bytes_per_second=40.3741G/s
Gather/double_coalesce_o/131072/2/manual_time        52636 ns        73786 ns        13041 bytes_per_second=74.2128G/s
Gather/double_coalesce_o/262144/2/manual_time        67713 ns        88321 ns        10415 bytes_per_second=115.376G/s
Gather/double_coalesce_o/524288/2/manual_time       114337 ns       130648 ns         6050 bytes_per_second=136.657G/s
Gather/double_coalesce_o/1048576/2/manual_time      304774 ns       319990 ns         2303 bytes_per_second=102.535G/s
Gather/double_coalesce_o/2097152/2/manual_time      990752 ns      1006723 ns          705 bytes_per_second=63.0834G/s
Gather/double_coalesce_o/4194304/2/manual_time     2411091 ns      2428085 ns          290 bytes_per_second=51.8438G/s
Gather/double_coalesce_o/8388608/2/manual_time     5237036 ns      5257208 ns          134 bytes_per_second=47.7369G/s
Gather/double_coalesce_o/16777216/2/manual_time   10986377 ns     11020105 ns           64 bytes_per_second=45.5109G/s
Gather/double_coalesce_o/33554432/2/manual_time   22654239 ns     22736868 ns           31 bytes_per_second=44.1418G/s
Gather/double_coalesce_o/67108864/2/manual_time   52437103 ns     52664220 ns           13 bytes_per_second=38.1409G/s
Gather/double_coalesce_o/1024/4/manual_time          58940 ns        80786 ns        10313 bytes_per_second=1060.4M/s
Gather/double_coalesce_o/2048/4/manual_time          59343 ns        80700 ns        11359 bytes_per_second=2.05703G/s
Gather/double_coalesce_o/4096/4/manual_time          59920 ns        81112 ns        11394 bytes_per_second=4.07446G/s
Gather/double_coalesce_o/8192/4/manual_time          61510 ns        81741 ns        10952 bytes_per_second=7.93829G/s
Gather/double_coalesce_o/16384/4/manual_time         60482 ns        81116 ns        11042 bytes_per_second=16.1464G/s
Gather/double_coalesce_o/32768/4/manual_time         60318 ns        81025 ns        11318 bytes_per_second=32.3803G/s
Gather/double_coalesce_o/65536/4/manual_time         61591 ns        81887 ns        11061 bytes_per_second=63.4226G/s
Gather/double_coalesce_o/131072/4/manual_time        71344 ns        91483 ns         9728 bytes_per_second=109.504G/s
Gather/double_coalesce_o/262144/4/manual_time       104217 ns       123578 ns         6795 bytes_per_second=149.927G/s
Gather/double_coalesce_o/524288/4/manual_time       185194 ns       201473 ns         3754 bytes_per_second=168.742G/s
Gather/double_coalesce_o/1048576/4/manual_time      569617 ns       585583 ns         1230 bytes_per_second=109.723G/s
Gather/double_coalesce_o/2097152/4/manual_time     1935376 ns      1952115 ns          362 bytes_per_second=64.5869G/s
Gather/double_coalesce_o/4194304/4/manual_time     4740904 ns      4760212 ns          148 bytes_per_second=52.7326G/s
Gather/double_coalesce_o/8388608/4/manual_time    10408969 ns     10440756 ns           67 bytes_per_second=48.0355G/s
Gather/double_coalesce_o/16777216/4/manual_time   21915651 ns     21996459 ns           32 bytes_per_second=45.6295G/s
Gather/double_coalesce_o/33554432/4/manual_time   45368122 ns     45647307 ns           15 bytes_per_second=44.0838G/s
Gather/double_coalesce_o/67108864/4/manual_time  104664717 ns    105639482 ns            7 bytes_per_second=38.2173G/s
Gather/double_coalesce_o/1024/8/manual_time          87414 ns       108018 ns         7646 bytes_per_second=1.39647G/s
Gather/double_coalesce_o/2048/8/manual_time          88322 ns       108726 ns         7767 bytes_per_second=2.76422G/s
Gather/double_coalesce_o/4096/8/manual_time          87780 ns       108169 ns         7844 bytes_per_second=5.56253G/s
Gather/double_coalesce_o/8192/8/manual_time          89813 ns       108522 ns         7814 bytes_per_second=10.8733G/s
Gather/double_coalesce_o/16384/8/manual_time         87753 ns       107744 ns         7797 bytes_per_second=22.257G/s
Gather/double_coalesce_o/32768/8/manual_time         87315 ns       107460 ns         7849 bytes_per_second=44.7375G/s
Gather/double_coalesce_o/65536/8/manual_time         93216 ns       112897 ns         7472 bytes_per_second=83.8108G/s
Gather/double_coalesce_o/131072/8/manual_time       112276 ns       132174 ns         6199 bytes_per_second=139.166G/s
Gather/double_coalesce_o/262144/8/manual_time       176019 ns       195611 ns         3953 bytes_per_second=177.538G/s
Gather/double_coalesce_o/524288/8/manual_time       327949 ns       344982 ns         2125 bytes_per_second=190.579G/s
Gather/double_coalesce_o/1048576/8/manual_time     1128786 ns      1110319 ns          617 bytes_per_second=110.738G/s
Gather/double_coalesce_o/2097152/8/manual_time     3840475 ns      3859390 ns          182 bytes_per_second=65.0961G/s
Gather/double_coalesce_o/4194304/8/manual_time     9457359 ns      9486382 ns           74 bytes_per_second=52.8689G/s
Gather/double_coalesce_o/8388608/8/manual_time    20773447 ns     20849079 ns           34 bytes_per_second=48.1384G/s
Gather/double_coalesce_o/16777216/8/manual_time   43776178 ns     44041521 ns           16 bytes_per_second=45.6869G/s
Gather/double_coalesce_o/33554432/8/manual_time   91154689 ns     92153943 ns            8 bytes_per_second=43.8815G/s
Gather/double_coalesce_o/67108864/8/manual_time  211007759 ns    216262385 ns            3 bytes_per_second=37.9133G/s
```
this refactor branch running on my machine:
```
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
----------------------------------------------------------------------------------------------------------
Benchmark                                                Time             CPU   Iterations UserCounters...
----------------------------------------------------------------------------------------------------------
Gather/double_coalesce_x/1024/1/manual_time          32756 ns        54235 ns        21621 bytes_per_second=477.008M/s
Gather/double_coalesce_x/2048/1/manual_time          31931 ns        53027 ns        21401 bytes_per_second=978.68M/s
Gather/double_coalesce_x/4096/1/manual_time          32244 ns        53317 ns        21694 bytes_per_second=1.89292G/s
Gather/double_coalesce_x/8192/1/manual_time          32011 ns        53113 ns        21490 bytes_per_second=3.8134G/s
Gather/double_coalesce_x/16384/1/manual_time         33213 ns        54768 ns        21215 bytes_per_second=7.35079G/s
Gather/double_coalesce_x/32768/1/manual_time         33227 ns        54483 ns        21054 bytes_per_second=14.6953G/s
Gather/double_coalesce_x/65536/1/manual_time         34605 ns        55782 ns        20350 bytes_per_second=28.2199G/s
Gather/double_coalesce_x/131072/1/manual_time        43031 ns        64184 ns        16362 bytes_per_second=45.3887G/s
Gather/double_coalesce_x/262144/1/manual_time        50156 ns        71650 ns        13957 bytes_per_second=77.8827G/s
Gather/double_coalesce_x/524288/1/manual_time        60239 ns        81013 ns        11417 bytes_per_second=129.692G/s
Gather/double_coalesce_x/1048576/1/manual_time       89404 ns       106303 ns         7819 bytes_per_second=174.769G/s
Gather/double_coalesce_x/2097152/1/manual_time      142549 ns       157994 ns         4888 bytes_per_second=219.223G/s
Gather/double_coalesce_x/4194304/1/manual_time      249084 ns       264851 ns         2807 bytes_per_second=250.92G/s
Gather/double_coalesce_x/8388608/1/manual_time      456665 ns       472610 ns         1535 bytes_per_second=273.724G/s
Gather/double_coalesce_x/16777216/1/manual_time     873582 ns       890108 ns          799 bytes_per_second=286.178G/s
Gather/double_coalesce_x/33554432/1/manual_time    1716411 ns      1727122 ns          409 bytes_per_second=291.306G/s
Gather/double_coalesce_x/67108864/1/manual_time    3377955 ns      3404098 ns          207 bytes_per_second=296.037G/s
Gather/double_coalesce_x/1024/2/manual_time          37986 ns        59572 ns        18371 bytes_per_second=822.669M/s
Gather/double_coalesce_x/2048/2/manual_time          37348 ns        58476 ns        18576 bytes_per_second=1.63424G/s
Gather/double_coalesce_x/4096/2/manual_time          37109 ns        58344 ns        18444 bytes_per_second=3.28949G/s
Gather/double_coalesce_x/8192/2/manual_time          43301 ns        64969 ns        16231 bytes_per_second=5.63827G/s
Gather/double_coalesce_x/16384/2/manual_time         44093 ns        65492 ns        15743 bytes_per_second=11.0739G/s
Gather/double_coalesce_x/32768/2/manual_time         44712 ns        65875 ns        15657 bytes_per_second=21.8414G/s
Gather/double_coalesce_x/65536/2/manual_time         46484 ns        68000 ns        15095 bytes_per_second=42.0172G/s
Gather/double_coalesce_x/131072/2/manual_time        51590 ns        72990 ns        13457 bytes_per_second=75.7178G/s
Gather/double_coalesce_x/262144/2/manual_time        62173 ns        82995 ns        11286 bytes_per_second=125.657G/s
Gather/double_coalesce_x/524288/2/manual_time        89165 ns       109784 ns         7399 bytes_per_second=175.237G/s
Gather/double_coalesce_x/1048576/2/manual_time      148271 ns       161082 ns         4828 bytes_per_second=210.762G/s
Gather/double_coalesce_x/2097152/2/manual_time      252931 ns       268430 ns         2760 bytes_per_second=247.103G/s
Gather/double_coalesce_x/4194304/2/manual_time      460179 ns       476131 ns         1519 bytes_per_second=271.634G/s
Gather/double_coalesce_x/8388608/2/manual_time      881967 ns       898993 ns          792 bytes_per_second=283.457G/s
Gather/double_coalesce_x/16777216/2/manual_time    1718687 ns      1737850 ns          407 bytes_per_second=290.92G/s
Gather/double_coalesce_x/33554432/2/manual_time    3387221 ns      3413174 ns          207 bytes_per_second=295.227G/s
Gather/double_coalesce_x/67108864/2/manual_time    6730877 ns      6785832 ns          104 bytes_per_second=297.138G/s
Gather/double_coalesce_x/1024/4/manual_time          59538 ns        81183 ns        11297 bytes_per_second=1049.74M/s
Gather/double_coalesce_x/2048/4/manual_time          58232 ns        79964 ns        11595 bytes_per_second=2.09626G/s
Gather/double_coalesce_x/4096/4/manual_time          59048 ns        80574 ns        11469 bytes_per_second=4.13459G/s
Gather/double_coalesce_x/8192/4/manual_time          58919 ns        80545 ns        11443 bytes_per_second=8.28736G/s
Gather/double_coalesce_x/16384/4/manual_time         59851 ns        81077 ns        10627 bytes_per_second=16.3165G/s
Gather/double_coalesce_x/32768/4/manual_time         59480 ns        80520 ns        11378 bytes_per_second=32.8368G/s
Gather/double_coalesce_x/65536/4/manual_time         59533 ns        81128 ns        11292 bytes_per_second=65.6153G/s
Gather/double_coalesce_x/131072/4/manual_time        65240 ns        85858 ns        10731 bytes_per_second=119.751G/s
Gather/double_coalesce_x/262144/4/manual_time        93088 ns       113851 ns         7472 bytes_per_second=167.852G/s
Gather/double_coalesce_x/524288/4/manual_time       148127 ns       168235 ns         4705 bytes_per_second=210.967G/s
Gather/double_coalesce_x/1048576/4/manual_time      252166 ns       268802 ns         2788 bytes_per_second=247.852G/s
Gather/double_coalesce_x/2097152/4/manual_time      472036 ns       487880 ns         1481 bytes_per_second=264.81G/s
Gather/double_coalesce_x/4194304/4/manual_time      888729 ns       905255 ns          788 bytes_per_second=281.301G/s
Gather/double_coalesce_x/8388608/4/manual_time     1725252 ns      1743641 ns          406 bytes_per_second=289.813G/s
Gather/double_coalesce_x/16777216/4/manual_time    3400040 ns      3425930 ns          206 bytes_per_second=294.114G/s
Gather/double_coalesce_x/33554432/4/manual_time    6756892 ns      6813065 ns          104 bytes_per_second=295.994G/s
Gather/double_coalesce_x/67108864/4/manual_time   13445374 ns     13616308 ns           52 bytes_per_second=297.5G/s
Gather/double_coalesce_x/1024/8/manual_time          88402 ns       107880 ns         7686 bytes_per_second=1.38086G/s
Gather/double_coalesce_x/2048/8/manual_time          86652 ns       107202 ns         7853 bytes_per_second=2.81748G/s
Gather/double_coalesce_x/4096/8/manual_time          88689 ns       108252 ns         7837 bytes_per_second=5.50552G/s
Gather/double_coalesce_x/8192/8/manual_time          87216 ns       107497 ns         7870 bytes_per_second=11.197G/s
Gather/double_coalesce_x/16384/8/manual_time         87085 ns       107272 ns         7424 bytes_per_second=22.4278G/s
Gather/double_coalesce_x/32768/8/manual_time         87174 ns       107581 ns         7896 bytes_per_second=44.81G/s
Gather/double_coalesce_x/65536/8/manual_time         87994 ns       108305 ns         7296 bytes_per_second=88.7844G/s
Gather/double_coalesce_x/131072/8/manual_time       100454 ns       120683 ns         6911 bytes_per_second=155.545G/s
Gather/double_coalesce_x/262144/8/manual_time       156682 ns       176690 ns         4448 bytes_per_second=199.448G/s
Gather/double_coalesce_x/524288/8/manual_time       264449 ns       281478 ns         2658 bytes_per_second=236.341G/s
Gather/double_coalesce_x/1048576/8/manual_time      466645 ns       484551 ns         1498 bytes_per_second=267.87G/s
Gather/double_coalesce_x/2097152/8/manual_time      907861 ns       924496 ns          772 bytes_per_second=275.372G/s
Gather/double_coalesce_x/4194304/8/manual_time     1740341 ns      1759066 ns          402 bytes_per_second=287.3G/s
Gather/double_coalesce_x/8388608/8/manual_time     3422898 ns      3450244 ns          205 bytes_per_second=292.15G/s
Gather/double_coalesce_x/16777216/8/manual_time    6764525 ns      6819753 ns          103 bytes_per_second=295.66G/s
Gather/double_coalesce_x/33554432/8/manual_time   13457101 ns     13627166 ns           52 bytes_per_second=297.241G/s
Gather/double_coalesce_x/67108864/8/manual_time   26843157 ns     27468544 ns           26 bytes_per_second=298.028G/s
Gather/double_coalesce_o/1024/1/manual_time          32968 ns        54274 ns        20839 bytes_per_second=473.95M/s
Gather/double_coalesce_o/2048/1/manual_time          32674 ns        53859 ns        21641 bytes_per_second=956.428M/s
Gather/double_coalesce_o/4096/1/manual_time          32911 ns        53930 ns        21395 bytes_per_second=1.85454G/s
Gather/double_coalesce_o/8192/1/manual_time          32717 ns        53780 ns        21330 bytes_per_second=3.73112G/s
Gather/double_coalesce_o/16384/1/manual_time         34216 ns        55399 ns        20797 bytes_per_second=7.13531G/s
Gather/double_coalesce_o/32768/1/manual_time         34706 ns        55879 ns        20058 bytes_per_second=14.0691G/s
Gather/double_coalesce_o/65536/1/manual_time         40131 ns        61190 ns        17764 bytes_per_second=24.3341G/s
Gather/double_coalesce_o/131072/1/manual_time        46458 ns        67464 ns        15099 bytes_per_second=42.0404G/s
Gather/double_coalesce_o/262144/1/manual_time        52177 ns        73408 ns        13326 bytes_per_second=74.8648G/s
Gather/double_coalesce_o/524288/1/manual_time        77188 ns        94100 ns         9005 bytes_per_second=101.214G/s
Gather/double_coalesce_o/1048576/1/manual_time      160409 ns       175847 ns         4357 bytes_per_second=97.4073G/s
Gather/double_coalesce_o/2097152/1/manual_time      529656 ns       534257 ns         1340 bytes_per_second=59.0006G/s
Gather/double_coalesce_o/4194304/1/manual_time     1214462 ns      1230827 ns          576 bytes_per_second=51.4631G/s
Gather/double_coalesce_o/8388608/1/manual_time     2643518 ns      2660729 ns          264 bytes_per_second=47.2855G/s
Gather/double_coalesce_o/16777216/1/manual_time    5512597 ns      5532578 ns          127 bytes_per_second=45.3507G/s
Gather/double_coalesce_o/33554432/1/manual_time   11297265 ns     11330132 ns           61 bytes_per_second=44.2585G/s
Gather/double_coalesce_o/67108864/1/manual_time   26212325 ns     26301827 ns           27 bytes_per_second=38.15G/s
Gather/double_coalesce_o/1024/2/manual_time          38475 ns        59930 ns        18246 bytes_per_second=812.215M/s
Gather/double_coalesce_o/2048/2/manual_time          38647 ns        59721 ns        17952 bytes_per_second=1.57928G/s
Gather/double_coalesce_o/4096/2/manual_time          38385 ns        59476 ns        18291 bytes_per_second=3.18017G/s
Gather/double_coalesce_o/8192/2/manual_time          43783 ns        65457 ns        16302 bytes_per_second=5.57612G/s
Gather/double_coalesce_o/16384/2/manual_time         45363 ns        66654 ns        15380 bytes_per_second=10.7638G/s
Gather/double_coalesce_o/32768/2/manual_time         46165 ns        67404 ns        14742 bytes_per_second=21.1539G/s
Gather/double_coalesce_o/65536/2/manual_time         49079 ns        70099 ns        14419 bytes_per_second=39.7955G/s
Gather/double_coalesce_o/131072/2/manual_time        52946 ns        74444 ns        12599 bytes_per_second=73.7786G/s
Gather/double_coalesce_o/262144/2/manual_time        66570 ns        87592 ns        10463 bytes_per_second=117.358G/s
Gather/double_coalesce_o/524288/2/manual_time       117299 ns       131624 ns         5838 bytes_per_second=133.206G/s
Gather/double_coalesce_o/1048576/2/manual_time      305033 ns       320550 ns         2295 bytes_per_second=102.448G/s
Gather/double_coalesce_o/2097152/2/manual_time     1007012 ns       992634 ns          697 bytes_per_second=62.0648G/s
Gather/double_coalesce_o/4194304/2/manual_time     2412089 ns      2429287 ns          290 bytes_per_second=51.8223G/s
Gather/double_coalesce_o/8388608/2/manual_time     5235452 ns      5255375 ns          134 bytes_per_second=47.7514G/s
Gather/double_coalesce_o/16777216/2/manual_time   10980486 ns     11012820 ns           64 bytes_per_second=45.5353G/s
Gather/double_coalesce_o/33554432/2/manual_time   22621070 ns     22702587 ns           31 bytes_per_second=44.2066G/s
Gather/double_coalesce_o/67108864/2/manual_time   52452352 ns     52771212 ns           13 bytes_per_second=38.1298G/s
Gather/double_coalesce_o/1024/4/manual_time          59685 ns        81269 ns        11450 bytes_per_second=1047.17M/s
Gather/double_coalesce_o/2048/4/manual_time          58818 ns        80508 ns        11112 bytes_per_second=2.0754G/s
Gather/double_coalesce_o/4096/4/manual_time          59452 ns        81088 ns        11373 bytes_per_second=4.1065G/s
Gather/double_coalesce_o/8192/4/manual_time          60155 ns        81147 ns        11332 bytes_per_second=8.11709G/s
Gather/double_coalesce_o/16384/4/manual_time         61332 ns        81601 ns        11238 bytes_per_second=15.9226G/s
Gather/double_coalesce_o/32768/4/manual_time         60314 ns        81111 ns        11438 bytes_per_second=32.3827G/s
Gather/double_coalesce_o/65536/4/manual_time         61961 ns        82152 ns        10633 bytes_per_second=63.0439G/s
Gather/double_coalesce_o/131072/4/manual_time        71112 ns        91723 ns         9747 bytes_per_second=109.862G/s
Gather/double_coalesce_o/262144/4/manual_time       106408 ns       126035 ns         6560 bytes_per_second=146.841G/s
Gather/double_coalesce_o/524288/4/manual_time       185630 ns       202217 ns         3762 bytes_per_second=168.346G/s
Gather/double_coalesce_o/1048576/4/manual_time      582296 ns       581011 ns         1225 bytes_per_second=107.334G/s
Gather/double_coalesce_o/2097152/4/manual_time     1945805 ns      1962818 ns          362 bytes_per_second=64.2408G/s
Gather/double_coalesce_o/4194304/4/manual_time     4765687 ns      4786018 ns          147 bytes_per_second=52.4583G/s
Gather/double_coalesce_o/8388608/4/manual_time    10414131 ns     10446114 ns           67 bytes_per_second=48.0117G/s
Gather/double_coalesce_o/16777216/4/manual_time   21893703 ns     21972086 ns           32 bytes_per_second=45.6752G/s
Gather/double_coalesce_o/33554432/4/manual_time   45287910 ns     45565447 ns           15 bytes_per_second=44.1619G/s
Gather/double_coalesce_o/67108864/4/manual_time  104961158 ns    106082828 ns            7 bytes_per_second=38.1093G/s
Gather/double_coalesce_o/1024/8/manual_time          87129 ns       107891 ns         7922 bytes_per_second=1.40102G/s
Gather/double_coalesce_o/2048/8/manual_time          88094 ns       108500 ns         7963 bytes_per_second=2.77137G/s
Gather/double_coalesce_o/4096/8/manual_time          89600 ns       108763 ns         7789 bytes_per_second=5.44957G/s
Gather/double_coalesce_o/8192/8/manual_time          87692 ns       107884 ns         7806 bytes_per_second=11.1362G/s
Gather/double_coalesce_o/16384/8/manual_time         86822 ns       107021 ns         7912 bytes_per_second=22.4957G/s
Gather/double_coalesce_o/32768/8/manual_time         89234 ns       108961 ns         7901 bytes_per_second=43.7754G/s
Gather/double_coalesce_o/65536/8/manual_time         93139 ns       113406 ns         7378 bytes_per_second=83.88G/s
Gather/double_coalesce_o/131072/8/manual_time       112236 ns       132192 ns         6189 bytes_per_second=139.216G/s
Gather/double_coalesce_o/262144/8/manual_time       176106 ns       195695 ns         3964 bytes_per_second=177.45G/s
Gather/double_coalesce_o/524288/8/manual_time       328329 ns       345556 ns         2070 bytes_per_second=190.358G/s
Gather/double_coalesce_o/1048576/8/manual_time     1116463 ns      1133580 ns          627 bytes_per_second=111.961G/s
Gather/double_coalesce_o/2097152/8/manual_time     3836531 ns      3855195 ns          182 bytes_per_second=65.163G/s
Gather/double_coalesce_o/4194304/8/manual_time     9455891 ns      9485865 ns           74 bytes_per_second=52.8771G/s
Gather/double_coalesce_o/8388608/8/manual_time    20862941 ns     20939933 ns           34 bytes_per_second=47.9319G/s
Gather/double_coalesce_o/16777216/8/manual_time   43803278 ns     44062787 ns           16 bytes_per_second=45.6587G/s
Gather/double_coalesce_o/33554432/8/manual_time   91493892 ns     92492610 ns            8 bytes_per_second=43.7188G/s
Gather/double_coalesce_o/67108864/8/manual_time  210335717 ns    215576331 ns            3 bytes_per_second=38.0344G/s
```

cwharris commented:
rerun tests

cwharris commented:
@jrhemstad Is that enough performance information?

jrhemstad commented:
> @jrhemstad Is that enough performance information? ^^

Yeah, these performance numbers are great. Glad to see that there wasn't any impact on performance.

cwharris commented:
@jrhemstad I think that makes sense, but I'm wondering if the iterators will introduce overhead or if they'll get optimized out in simpler cases.

jrhemstad commented:
> @jrhemstad I think that makes sense, but I'm wondering if the iterators will introduce overhead or if they'll get optimized out in simpler cases.

They should definitely get optimized out/inlined. That's the beauty of iterators. Since they are templated, they are known at compile time, and can therefore be inlined and any extra overhead is boiled away by the compiler.

cwharris commented:
@jrhemstad Most use-cases for `valid_if` use zero-based counting iterators rather than anything fancy, with the exception of `bools_to_mask`, which uses an iterator of `experimental::bool8`, but this could have easily used a zero-based counting iterator with:
```c++
auto pred = __device__ [begin=device_input.begin()](size_type idx){ return begin + idx; }
```

`select_bitmask_kernel` determines the `col` / `row`, and passes that to the predicate. The predicate then looks-up multiple values based on these two parameters. To do the same thing with iterators would require two counting iterators and two transform iterators, each creating a `pair`, one of which would contain the original row id.

I think the intent is clearer if that is delegated to the selector. I choose to call it "selector", because predicate implies it's doing a simplish comparison, but in reality it could do be looking across multiple device inputs (`&`-ing each input column's validity, for instance).

Because the algorithm is conceptually dependent on a two dimensional data structure (the table), I'm not sure it makes sense to move it to `valid_if`. If `valid_if` were intended to handle multi-dimensional inputs / outputs, and parallelise them appropriately, it would make sense, but I think that's more work than I'm wanting to accomplish right now.

If you don't feel to strongly about it, I'd rather keep the current implementation and begin using it in other tasks.

jrhemstad commented:
> Most use-cases for `valid_if` use zero-based counting iterators rather than anything fancy

Perhaps, but it's just good algorithm design to make it more flexible when you can. For example, no STL algorithm implicitly uses the equivalent of a "counting iterator", you can always specify the range over which a predicate is applied.

The fact that the usescases are simple now doesn't mean we won't encounter a case in the future where we'll be glad that it was made to be more generic. 

> To do the same thing with iterators would require two counting iterators and two transform iterators, each creating a pair, one of which would contain the original row id.

I'm not following. I gave outlined how the existing behavior would be preserved and it didn't require any transform iterators or `pairs`.

> predicate implies it's doing a simplish comparison

"predicate" is the idiomatic name for a function that returns `true` or `false`. It can be as simple or complex as a person wants it to be. In this case, we have a function that takes two inputs (binary) and returns `true/false` based on those inputs, hence "binary predicate". 

> If valid_if were intended to handle multi-dimensional inputs / outputs

The current `valid_if` isn't, but your new one would be. I'm not suggesting modifying the old `valid_if`, but instead adding a new `valid_if_n` (or similar name). 

> If you don't feel too strongly about it, I'd rather keep the current implementation and begin using it in other tasks.

Think about what your function does in one sentence "Returns a set of bitmasks where the state of each bit is determined by a binary predicate". That sounds a lot like `valid_if`, "Returns a single bitmask where the state of each bit is determined by a unary predicate.".

So how about a compromise. Move the implementation to `valid_if.cuh`. Keep the "implicit counting iterator", but rename "selector" to "BinaryPredicate p". That'll make it a bit more idiomatic and descriptive (`select_bitmask` isn't quite self descriptive), and set us up to add input iterators at a later time if we so choose.

cwharris commented:
@jrhemstad I think you made the stronger argument here, esp. with regards to consistency with existing stl apis. Moved to `valid_if.cuh`, renamed to `valid_if_n_kernel` (I'm open to different names), made it iterator based, and updated naming to reflect binary predicate, rather than selector.

jrhemstad commented on cpp/include/cudf/detail/valid_if.cuh:
```diff
@@ -115,6 +116,71 @@ std::pair<rmm::device_buffer, size_type> valid_if(
+ *
+ * @note If any mask is `nullptr`, that mask will be ignored.
+ *
+ * @param get_validity    Function used to obtain validity for each bit.
```
Update naming.

jrhemstad commented on cpp/include/cudf/detail/valid_if.cuh:
```diff
@@ -115,6 +116,71 @@ std::pair<rmm::device_buffer, size_type> valid_if(
+
+/**----------------------------------------------------------------------------*
+ * @brief For each mask, assigns each bit to a value obtained by calling the
+ *        predicate.
```
Docs could use a little expounding. 

```
Populates a set of bitmasks by applying a binary predicate to two input ranges.

Given a set of bitmasks, `masks`, the state of bit `j` in mask `i` is determined by `p( *(begin1 + i), *(begin2 + j))`.
If the predicate evaluates to `true`, the bit is set to `1`. If false, set to `0`.
```

trevorsm7 commented on cpp/include/cudf/detail/gather.cuh:
```diff
@@ -1,3 +1,19 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
```
```suggestion
 * Copyright (c) 2019-2020, NVIDIA CORPORATION.
```

trevorsm7 commented on cpp/include/cudf/detail/gather.cuh:
```diff
@@ -1,3 +1,19 @@
+ * limitations under the License.
+ */
+
 #include <cudf/cudf.h>
```
Is it possible to remove this header?
```suggestion
```

rgsl888prabhu commented:
rerun tests

kkraus14 commented:
rerun tests

cwharris commented:
rerun tests
